posts/fantastic-simprocs.md

```diff
@@ -0,0 +1,173 @@
+---
+author: 'Yaël Dillies, Paul Lezeau'
+category: 'Metaprogramming'
+date: 2025-03-29 12:33:00 UTC+00:00
+description: 'Brief introduction to simprocs and what they are useful for'
+has_math: true
+link: ''
+slug: fantastic-simprocs
+tags: 'simp, simproc, meta'
+title: Fantastic Simprocs and Where to Find Them.
+type: text
+---
+
+Lean core added support for custom simplification procedures, aka *simprocs*.
+This blog post is the first in a series of two aimed at explaining what a simproc is, what kind of problems can be solved with simprocs, and what tools we have to write them.
+
+<!-- TEASER_END -->
+
+This post describes purely informally what simprocs are and do.
+The second post will be a walkthrough to writing a simple simproc in three different ways.
+
+# What is a simproc
+
+To understand what a simproc is and how it works, we will first explain how `simp` works.
+Then we will give some examples and non-examples of simprocs as well as pointers to analogous concepts in other theorem provers.
+
+## How simp works
+
+On an informal level, `simp` can be seen as recursively traversing the expression `e` to be simplified from the outside-in, each time checking whether the expression matches the left hand side `LHS` of some simp lemma of the form `LHS = RHS`, and replacing `LHS` by `RHS` in the expression if it does match.
```

Comment:
This appears inaccurate: note that the Lean reference manual says this: "The [simp](https://lean-lang.org/doc/reference/latest/Tactic-Proofs/Tactic-Reference/#simp) tactics primarily rewrite from the inside out."

Reply:
This is the same phrasing issue as here: https://github.com/leanprover-community/blog/pull/97#discussion_r2040673512

namely simp *traverses* from the outside-in, and *rewrites* from the inside-out

Reply:
Right, and I think that since the default and common case is postprocedures, it makes more sense for the informal description to talk about that case instead.

Reply:
Maybe it's worth mentioning also that any lemma (and simproc) can be set to act during the outside-in part, by adding the `\d` arrow in front of the identifier.

Reply:
I've also pushed some local changes that I had forgotten to add, which remove the "outside-in" from this sentence and give a tiny bit more information about how simp traverses expressions.